atlasdb-cassandra-integration-tests/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraTimestampBoundStoreTest.java

```diff
@@ -0,0 +1,199 @@
+    }
+
+    @Test
+    public void canGetOldFormatOnStartup() {
+        insertTimestampOld(OFFSET);
+        assertBoundInDbIsEqualTo(OFFSET);
+    }
+
+    @Test
+    public void getOldFormatThrowsAfterStore() {
+        store.getUpperLimit();
+        store.storeUpperLimit(SECOND_OFFSET);
+        insertTimestampOld(OFFSET);
+        assertThatThrownBy(() -> store.getUpperLimit()).isExactlyInstanceOf(MultipleRunningTimestampServiceError.class);
+    }
+
+    @Test
+    public void storeWithWrongTimestampCorrectIdSucceeds() {
+        long limit = store.getUpperLimit();
+        insertTimestampWithCorrectId(limit + OFFSET);
+        store.storeUpperLimit(limit + SECOND_OFFSET);
+        assertBoundInDbIsEqualTo(limit + SECOND_OFFSET);
+    }
+
+    @Test
+    public void storeWithWrongTimestampNoIdThrows() {
+        long limit = store.getUpperLimit();
+        insertTimestampOld(limit + OFFSET);
+        assertThatStoreUpperLimitThrows(limit + SECOND_OFFSET);
+    }
+
+    @Test
+    public void storeWithWrongTimestampWrongIdThrows() {
+        long limit = store.getUpperLimit();
+        insertTimestampWithFakeId(limit + OFFSET);
+        assertThatStoreUpperLimitThrows(limit + SECOND_OFFSET);
+    }
+
+    @Test
+    public void storeWithRightTimestampNoIdSucceedsFirstTime() {
+        insertTimestampOld(OFFSET);
+        assertThat(store.getUpperLimit()).isEqualTo(OFFSET);
+        store.storeUpperLimit(SECOND_OFFSET);
+        assertBoundInDbIsEqualTo(SECOND_OFFSET);
+    }
+
+    @Test
+    public void storeWithRightTimestampNoIdFailsSecondTime() {
+        insertTimestampOld(OFFSET);
+        store.getUpperLimit();
+        store.storeUpperLimit(SECOND_OFFSET);
+        insertTimestampOld(OFFSET);
+        assertThatStoreUpperLimitThrows(SECOND_OFFSET);
+    }
+
+    @Test
+    public void storeWithRightTimestampWrongIdSucceedsFirstTime() {
+        insertTimestampWithFakeId(OFFSET);
+        assertThat(store.getUpperLimit()).isEqualTo(OFFSET);
+        store.storeUpperLimit(SECOND_OFFSET);
+        assertBoundInDbIsEqualTo(SECOND_OFFSET);
+    }
+
+    @Test
+    public void storeWithRightTimestampWrongIdFailsSecondTime() {
+        insertTimestampWithFakeId(OFFSET);
+        store.getUpperLimit();
+        store.storeUpperLimit(SECOND_OFFSET);
+        insertTimestampWithFakeId(OFFSET);
+        assertThatStoreUpperLimitThrows(SECOND_OFFSET);
+    }
+
+    @Test
+    public void valueBelowEightBytesThrows() {
+        setTimestampTableValueTo(PtBytes.toBytes("1"));
+        assertThatThrownBy(() -> store.getUpperLimit()).isExactlyInstanceOf(IllegalArgumentException.class);
+    }
+
+    @Test
+    public void invalidValueAboveEightBytesThrows() {
+        setTimestampTableValueTo(PtBytes.toBytes("123456789"));
+        assertThatThrownBy(() -> store.getUpperLimit()).isExactlyInstanceOf(IllegalArgumentException.class);
+    }
```

Comment:
nit: The other "edge" case here would be something >=25 bytes i.e. longer than a UUID + a long; maybe we could add a test for that. Feel free to push back.